### ISSUE TITLE
[LOW] [Splits] [$500] IOU - Bold, unread DMs become unbold without opening when app is opened in desktop and web

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#35652]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v.1.3.45-5
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:**

## Action Performed:
1. Launch desktop app.
2. Create a workspace.
3. Invite a few users with no prior chat to the workspace.
4. Go to #announce > + > Split bill.
5. Create a bill split with all members.
6. Note that 1:1 DMs in LHN are bold and do not turn unbold.
7. Open web app and log in with the same account as desktop app.
8. Repeat Step 2 to 5 on desktop app with web app open at the same time.

## Expected Result:
1:1 DMs in LHN will remain bold and do not turn unbold.

## Actual Result:
1:1 DMs in LHN are bold first and then turn unbold. This issue only occurs when desktop app and web app are opened and logged in to the same account at the same time.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/6d443df7-8404-40f5-b2c9-04317266888f

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018bb1604c465cc6ec</li>
        <li>Upwork Job ID: 1753370840941113344</li>
        <li>Last Price Increase: 2024-03-16</li>
    </ul>
</details>

**Comment by izarutskaya** [comment]
We think that this bug might be related to #vip-split-p2p-chat-groups

**Comment by yoyumiracle** [comment]
Contributor details
Your Expensify account email: [peraivan10@gmail.com](mailto:peraivan10@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~010e420e4da95306da

It seems problem occured in code below.

OptionRowLHNData.tsx file
const optionItem = useMemo(() => {
        // Note: ideally we'd have this as a dependent selector in onyx!
        const item = SidebarUtils.getOptionData({
            report: fullReport,
            reportActions,
            personalDetails,
            preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
            policy,
            parentReportAction,
            hasViolations: !!hasViolations,
        });
        if (deepEqual(item, optionItemRef.current)) {
            return optionItemRef.current;
        }

        optionItemRef.current = item;

        return item;
        // Listen parentReportAction to update title of thread report when parentReportAction changed
        // Listen to transaction to update title of transaction report when transaction changed
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [fullReport, linkedTransaction, reportActions, personalDetails, preferredLocale, policy, parentReportAction, transaction, transactionViolations, canUseViolations]);

Solution: 
OptionItem.isUnread variable seems to be controlled in local state. Desktop app Value and MobileApp value maybe conflict to this variable.

**Comment by yoyumiracle** [comment]
Contributor details
Your Expensify account email: [peraivan10@gmail.com](mailto:peraivan10@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~010e420e4da95306da

It seems problem occured in code below.

OptionRowLHNData.tsx file
const optionItem = useMemo(() => {
// Note: ideally we'd have this as a dependent selector in onyx!
const item = SidebarUtils.getOptionData({
report: fullReport,
reportActions,
personalDetails,
preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
policy,
parentReportAction,
hasViolations: !!hasViolations,
});
if (deepEqual(item, optionItemRef.current)) {
return optionItemRef.current;
}

    optionItemRef.current = item;

    return item;
    // Listen parentReportAction to update title of thread report when parentReportAction changed
    // Listen to transaction to update title of transaction report when transaction changed
    // eslint-disable-next-line react-hooks/exhaustive-deps
}, [fullReport, linkedTransaction, reportActions, personalDetails, preferredLocale, policy, parentReportAction, transaction, transactionViolations, canUseViolations]);

Solution
OptionItem.isUnread variable seems to be controlled in local state. Desktop app Value and MobileApp value maybe conflict to this variable.

**Comment by yoyumiracle** [comment]
Contributor details
Your Expensify account email: [peraivan10@gmail.com](mailto:peraivan10@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~010e420e4da95306da

const optionItem = useMemo(() => {
        // Note: ideally we'd have this as a dependent selector in onyx!
        const item = SidebarUtils.getOptionData({
            report: fullReport,
            reportActions,
            personalDetails,
            preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
            policy,
            parentReportAction,
            hasViolations: !!hasViolations,
        });
        if (deepEqual(item, optionItemRef.current)) {
            return optionItemRef.current;
        }

        optionItemRef.current = item;

        return item;
        // Listen parentReportAction to update title of thread report when parentReportAction changed
        // Listen to transaction to update title of transaction report when transaction changed
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [fullReport, linkedTransaction, reportActions, personalDetails, preferredLocale, policy, parentReportAction, transaction, transactionViolations, canUseViolations]);

Solution
OptionItem.isUnread variable seems to be controlled in local state. Desktop app Value and MobileApp value maybe conflict to this variable.

**Comment by yoyumiracle** [comment]
Contributor details
Your Expensify account email: [peraivan10@gmail.com](mailto:peraivan10@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~010e420e4da95306da

Solution
OptionItem.isUnread variable seems to be controlled in local state. Desktop app Value and MobileApp value maybe conflict to this variable.

**Comment by yoyumiracle** [comment]
const optionItem = useMemo(() => {
// Note: ideally we'd have this as a dependent selector in onyx!
const item = SidebarUtils.getOptionData({
report: fullReport,
reportActions,
personalDetails,
preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
policy,
parentReportAction,
hasViolations: !!hasViolations,
});
if (deepEqual(item, optionItemRef.current)) {
return optionItemRef.current;
}

Solution
OptionItem.isUnread variable seems to be controlled in local state. Desktop app Value and MobileApp value maybe conflict to this variable.

**Comment by Santhosh-Sellavel** [comment]
@yoyumiracle Go through our [contributingGuides](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) carefully and write a formal proposal using [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md)

**Comment by yoyumiracle** [comment]
The problem is when users logged in Dapp(Desktop App) and Mapp(Mobile App) simultaneously the LHN bold turn to unbold.

I think root problem is occurred from the logic like below.
Dapp and Mapp should affect to OptionItem.isUnread variable at the same time.
According code is

const optionItem = useMemo(() => {
// Note: ideally we'd have this as a dependent selector in onyx!
const item = SidebarUtils.getOptionData({
report: fullReport,
reportActions,
personalDetails,
preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
policy,
parentReportAction,
hasViolations: !!hasViolations,
});
if (deepEqual(item, optionItemRef.current)) {
return optionItemRef.current;
}

Suggest solution is control the OptionItem.isUnread's value by useState not useMemo's return value. useMemo's return value must be used to initialize the OptionItem.isUnread's value.

**Comment by stevechainer** [comment]
Well done.
peraivan10@gmail.com is a perfect solution.
I thought it like him/her. 
I am shame to be a bit late, but he is same as me.

**Comment by Santhosh-Sellavel** [comment]
> @yoyumiracle Go through our [contributingGuides](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) carefully and write a formal proposal using [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md)

@yoyumiracle Had a chance to go through the guides. Can you submit an updated proposal using [proposal template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md)
Refer to the past [issues](https://github.com/Expensify/App/issues?q=label%3A%22Awaiting+Payment%22+is%3Aclosed) to understand what we expect thanks!

**Comment by yoyumiracle** [comment]
@Santhosh-Sellavel May I have an example of proposal? I read proposal template but it's not clear to me. Thanks.

**Comment by yoyumiracle** [comment]
## Proposal

### 1:1 DMs in LHN are bold first and then turn unbold. This issue only occurs when desktop app and web app are opened and logged in to the same account at the same time.

### By deep look in code "1:1 DMs in LHN" bold control variable is "OptionItem.isUnread". Seems to be it conflicted when desktop app and web app are opened and logged in to the same account simultaneously.

### I suggest that control the value on local state. It would be updated by Server when action performed. Server update value must be initialized when page loading and then it must be controlled by useState.

**Comment by yoyumiracle** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
1:1 DMs in LHN are bold first and then turn unbold. This issue only occurs when desktop app and web app are opened and logged in to the same account at the same time.

### What is the root cause of that problem?
By deep look in code "1:1 DMs in LHN" bold control variable is "OptionItem.isUnread". Seems to be it conflicted when desktop app and web app are opened and logged in to the same account simultaneously.

### What changes do you think we should make in order to solve the problem?
I suggest that control the value on local state. It would be updated by Server when action performed. Server update value must be initialized when page loading and then it must be controlled by useState.

**Comment by Santhosh-Sellavel** [comment]
> Seems to be conflicted when the desktop app and web app are opened and logged in to the same account simultaneously.

Can you elaborate on this?

> I suggest that control the value on local state. It would be updated by Server when action performed. Server update value must be initialized when page loading and then it must be controlled by useState.

Can you rephrase this and share more implementation details?

**Comment by yoyumiracle** [comment]
##Can you elaborate on this?
LHN's bold property is only controlled by optionItem.isUnread variable do you agree with it?
Once it turned bold, it will be never change until backend data changed.
This means there maybe conflict status between Desktop App and Web App.

## Can you rephrase this and share more implementation details?
Following code the optionItem.isUnread variable is changed in useMemo hook.
If we use useState hook to control optionItem.isUnread in frontend status and send variable data to server only It will never affected by Desktop or Web app.
To synchronous status please init optionItem.isUnread variable once on loading page.

**Comment by yoyumiracle** [comment]
### LHN Component
Analysis of code, I guess that LHN component is OptionRowLHN.tsx

### DisplayName Bold control logic
Look into OpetionRowLHN code displayname's style is controlled by textUnreadStyle and textStyle.
If it's unreaded textUnreadStyle applied otherwise textStyle is applied.

### textUnreadStyle
. const textUnreadStyle = optionItem?.isUnread && optionItem.notificationPreference !== CONST.REPORT.NOTIFICATION_PREFERENCE.MUTE ? [textStyle, styles.sidebarLinkTextBold] : [textStyle];
    const displayNameStyle = [styles.optionDisplayName, styles.optionDisplayNameCompact, styles.pre, textUnreadStyle, style];

### Suggest Method
optionItem.isUnread is controlled in OptionLowLHNData.tsx file's useMemo hook
It seems useMemo called twice with Desktop App and Web App simultaneously. So it turn unbold when login Desktop App and Web App as same user at same time.
So I suggest that control optionItem.isUnread variable individually at frontend and send status to server only.
To sync variable initialize with optionItem.isUnread value with server status only once at loading page

I hope it would help you solve the issue. Thanks.

**Comment by Santhosh-Sellavel** [comment]
@mallenexpensify
Sorry, somehow this one got missed from my radar, Unassigning it as am planning for OOO, please assign a new C+ to move this forward, thanks!

**Comment by mananjadhav** [comment]
@yoyumiracle I just took over and I can see multiple comments from you. Can you post one final proposal in the suggested format for me to review?

**Comment by yoyumiracle** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
1:1 DMs in LHN are bold first and then turn unbold. This issue only occurs when desktop app and web app are opened and logged in to the same account at the same time.

### What is the root cause of that problem?
Desktop App and Web App affect state of LHN's display name simultaneously.

### What changes do you think we should make in order to solve the problem?
 ### 1. LHN's display name element's style control.
#### 1-1. I think this code controls the LHN's display name element's style.
            const textUnreadStyle = optionItem?.isUnread && optionItem.notificationPreference !== 
            CONST.REPORT.NOTIFICATION_PREFERENCE.MUTE ? [textStyle, styles.sidebarLinkTextBold] : [textStyle];
     
#### 1-2. I suggest to change code like this.
            const textUnreadStyle = optionItem?.isUnread || optionItem.notificationPreference !== 
            CONST.REPORT.NOTIFICATION_PREFERENCE.MUTE ? [textStyle, styles.sidebarLinkTextBold] : [textStyle];

 ### 2. OptionItem.isUnread variable value control.
If first method doesn't work I think we have to control optionItem.isUnread value locally. This means that optionItem.isUnread value is affected somehow by Desktop App.
#### My opinion is like this.
Follow the optionItem.isUnread variable management flow, it is changed in useMemo hook in OptionRowLHNData.tsx.
2-1. I think optionItem.isUnread's value have to be initialised when page loading to sync with Desktop App or Server. 

       useEffect(..., [])
2-2. Then it's value must be controlled by useState hook independantly instead of useMemo and if value changing occured by locally then send it's value to server.

Code below is useMemo code

     const optionItem = useMemo(() => {
            // Note: ideally we'd have this as a dependent selector in onyx!
            const item = SidebarUtils.getOptionData({
                report: fullReport,
                reportActions,
                personalDetails,
                preferredLocale: preferredLocale ?? CONST.LOCALES.DEFAULT,
                policy,
                parentReportAction,
                hasViolations: !!hasViolations,
            });
            if (deepEqual(item, optionItemRef.current)) {
                return optionItemRef.current;
            }
    
            optionItemRef.current = item;
    
            return item;
            // Listen parentReportAction to update title of thread report when parentReportAction changed
            // Listen to transaction to update title of transaction report when transaction changed
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [fullReport, linkedTransaction, reportActions, personalDetails, preferredLocale, policy, parentReportAction, transaction, transactionViolations, canUseViolations]);

It's only my opinion and I haven't test it because of my environment.
Thanks

**Comment by yoyumiracle** [comment]
Hi @mananjadhav .
Is the proposal right or not? 
As I said in comment, I couldn't do testing on it. I wonder if you could tell me if it's gone through wrong way. 
I want to know that the code I posted on comment was really for LHN.

**Comment by mallenexpensify** [comment]
@mananjadhav 👀 please on @yoyumiracle 's proposal above.  Thx

**Comment by mananjadhav** [comment]
I'll review this today.

**Comment by mananjadhav** [comment]
@yoyumiracle It is not clear from the proposal what is the root cause of the problem? I also think a lot of the proposal is based on the assumption and now actual analysis. `This means that optionItem.isUnread value is affected somehow by Desktop App.`

I tried to understand the change in `Code below is useMemo code` but I can't. Also I would recommend don't paste large pieces of code.

> It's only my opinion and I haven't test it because of my environment.

I am not sure if I can be of any help here.  But the you can refer to [this guide](https://github.com/Expensify/App/blob/main/contributingGuides/TESTING_MACOS_AND_IOS.md) on how to test for Mac. I don't expect you to use paid services to post the proposals, but I am not confident if we're addressing the right root cause. You could attempt to test on Web + android.


Still open for better proposals.

**Comment by mallenexpensify** [comment]
Throwing `retest-weekly` on here to ensure it's still happening

**Comment by yoyumiracle** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
1:1 DMs in LHN are bold first and then turn unbold. This issue only occurs when desktop app and web app are opened and logged in to the same account at the same time.

### What is the root cause of that problem?
There's no synchronous with Desktop App and Web App.
I think the main point of this issue is that if we invite members to work in Desktop App then it's status must be applied to Web app at same time. But from the video I can't detect it. 

### What changes do you think we should make in order to solve the problem?
I think there's two way to solve problem.
. First is that make synchronous with Desktop App and Web App. If change state on one's account within one platform that must be applied to other platform at same time.
. Second, Just think about why one account could be log in each platform simultaneously? Is this logic right?

Thanks.

**Comment by mallenexpensify** [comment]
@mananjadhav can you please review the above proposal? Thx

**Comment by mallenexpensify** [comment]
I believe @mananjadhav 's been sick the past few days. 
@mananjadhav , please review when you can, thanks.

**Comment by mananjadhav** [comment]
Thanks for the proposal @yoyumiracle. I still think we are not there yet in terms of the bug/proposal.

>  Just think about why one account could be log in each platform simultaneously? Is this logic right?

You could say not every time, but why wouldn't one be logged in the platforms? I don't think we should be putting any such restrictions.

Still open for proposals.

**Comment by yoyumiracle** [comment]
@mananjadhav  Thanks for your reply.
I think that If a user who logged in Web app and Desktop App simultaneously create workspace in web app than at that time Desktop App must show same result of LHN's list on Web app. Isn't it?
It means that there would be mistake on backend. It's my opinion.
Thanks.

**Comment by mallenexpensify** [comment]
Stepping back a bit.... 
> when app is opened in desktop and web

1. I think this is likely rare (but not _impossible_, internally I have both open often)
2. If there's a bug with a report type being marked as read/unread (unbold/bold), is this just occurring for this one instance?  I would think there _might_ be others and, if so, it'd be great to combine them into one fix.  

@mananjadhav , what do you think?
I'm not raising the price because this is an edge case and priority is 'low'.

**Comment by mananjadhav** [comment]
I agree with the thoughts @mallenexpensify. I do think one could have both the apps open. Yes it's an edge case, so I don't think a price bump is required as of now.
